File: algorithms/21_scala/strings/Manacher.scala
object Manacher {
  def manacher(s: String): Int = {
    val t = "#" + s.mkString("#") + "#"
    val n = t.length
    val p = new Array[Int](n)
    var c = 0
    var r = 0
    for (i <- 0 until n) {
      if (i < r) p(i) = math.min(r - i, p(2 * c - i))
      while (i + p(i) + 1 < n && i - p(i) - 1 >= 0 &&
             t.charAt(i + p(i) + 1) == t.charAt(i - p(i) - 1))
        p(i) += 1
      if (i + p(i) > r) { c = i; r = i + p(i) }
    }
    p.max
  }

  def main(args: Array[String]): Unit = {
    assert(manacher("abba") == 4, "manacher even wrong")
    assert(manacher("racecar") == 7, "manacher odd wrong")
    println("[Scala Manacher] Longest palindromic substring verified")
  }
}

File: algorithms/19_lua/strings/manacher.lua
local M = {}

function M.manacher(s)
    local t = "#"
    for i = 1, #s do
        t = t .. s:sub(i, i) .. "#"
    end
    local n = #t
    local p = {}
    for i = 1, n do p[i] = 0 end
    local c, r = 0, 0
    for i = 1, n do
        if i <= r then p[i] = math.min(r - i, p[2 * c - i]) end
        while i + p[i] + 1 <= n and i - p[i] - 1 >= 1 and
              string.byte(t, i + p[i] + 1) == string.byte(t, i - p[i] - 1) do
            p[i] = p[i] + 1
        end
        if i + p[i] > r then
            c = i
            r = i + p[i]
        end
    end
    local best = 0
    for i = 1, n do
        if p[i] > best then best = p[i] end
    end
    return best
end

if not pcall(debug.getlocal, 4, 1) then
    assert(M.manacher("abba") == 4, "manacher even wrong")
    assert(M.manacher("racecar") == 7, "manacher odd wrong")
    print("[Lua Manacher] Longest palindromic substring verified")
end

return M
